drivers/net/w5500/w5500.c

```diff
@@ -0,0 +1,1099 @@
+}
+
+/***************************************************************************//**
+ * @brief Connect to a remote host (TCP client mode)
+ *
+ * @param dev     - Device descriptor
+ * @param sock_id - Socket number (0-7)
+ * @param addr    - Socket address structure containing remote IP and port
+ *
+ * @return 0 on success, negative error code otherwise
+*******************************************************************************/
+int w5500_socket_connect(struct w5500_dev *dev, uint8_t sock_id,
+			 struct w5500_socket_address *addr)
+{
+	int ret;
+	uint8_t status;
+
+	if (sock_id > W5500_MAX_SOCK_NUMBER)
+		return -EINVAL;
+
+	if (!addr)
+		return -EINVAL;
+
+	ret = w5500_socket_reg_read(dev, sock_id, W5500_Sn_SR, &status, 1);
+	if (ret)
+		return ret;
+
+	if (status != W5500_Sn_SR_INIT)
+		return -EINVAL;
+
+	ret = w5500_socket_reg_write(dev, sock_id, W5500_Sn_DIPR, addr->ip, 4);
+	if (ret)
+		return ret;
+
+	ret = w5500_socket_reg_write(dev, sock_id, W5500_Sn_DPORT, addr->port, 2);
+	if (ret)
+		return ret;
+
+	return w5500_socket_command_write(dev, sock_id, W5500_Sn_CR_CONNECT);
+}
+
+/***************************************************************************//**
+ * @brief Disconnect a TCP connection
+ *
+ * @param dev     - Device descriptor
+ * @param sock_id - Socket number (0-7)
+ *
+ * @return 0 on success, -ENOTCONN if not connected, other negative error codes
+*******************************************************************************/
+int w5500_socket_disconnect(struct w5500_dev *dev, uint8_t sock_id)
+{
+	int ret;
+	uint8_t status;
+
+	if (sock_id > W5500_MAX_SOCK_NUMBER)
+		return -EINVAL;
+
+	ret = w5500_socket_reg_read(dev, sock_id, W5500_Sn_SR, &status, 1);
+	if (ret)
+		return ret;
+
+	if (status != W5500_Sn_SR_ESTABLISHED &&
+	    status != W5500_Sn_SR_CLOSE_WAIT)
+		return -ENOTCONN;
+
+	return w5500_socket_command_write(dev, sock_id, W5500_Sn_CR_DISCON);
+}
+
+/***************************************************************************//**
+ * @brief Send data through a socket
+ *
+ * @param dev     - Device descriptor
+ * @param sock_id - Socket number (0-7)
+ * @param buf     - Data buffer to send
+ * @param len     - Data length
+ *
+ * @return Number of bytes sent on success, negative error code otherwise
+*******************************************************************************/
+int w5500_socket_send(struct w5500_dev *dev, uint8_t sock_id, const void *buf,
+		      uint16_t len)
+{
+	int ret;
+	uint16_t free_size = 0;
+	uint16_t wr_ptr;
+	uint8_t wr_ptr_buf[2];
+	uint8_t status;
+	uint16_t total_sent = 0;
+	uint16_t chunk_size;
+	const uint8_t *data = (const uint8_t *)buf;
+
+	if (sock_id > W5500_MAX_SOCK_NUMBER)
+		return -EINVAL;
+
+	ret = w5500_socket_reg_read(dev, sock_id, W5500_Sn_SR, &status, 1);
+	if (ret)
+		return ret;
+
+	if (status != W5500_Sn_SR_ESTABLISHED &&
+	    status != W5500_Sn_SR_UDP &&
+	    status != W5500_Sn_SR_MACRAW)
+		return -ENOTCONN;
+
+	while (total_sent < len) {
```

Comment:
Would it be possible to remain stuck in this loop (e.g the other side disconnects while the buffer is full)? If so, I'd add a timeout.